cluster-autoscaler/provisioningrequest/checkcapacity/processor.go

```diff
@@ -0,0 +1,79 @@
+/*
+Copyright 2024 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package checkcapacity
+
+import (
+	"fmt"
+	"time"
+
+	apimeta "k8s.io/apimachinery/pkg/api/meta"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/autoscaler/cluster-autoscaler/provisioningrequest/apis/autoscaling.x-k8s.io/v1beta1"
+	"k8s.io/autoscaler/cluster-autoscaler/provisioningrequest/provreqclient"
+	"k8s.io/client-go/rest"
+)
+
+const (
+	defaultReservationTime = 10 * time.Minute
+	defaultExpirationTime  = 7 * 24 * time.Hour // 7 days
+)
+
+type checkCapacityProcessor struct {
+	client provisioningRequestClient
+	now    func() time.Time
+}
+
+// NewCheckCapacityProcessor return ProvisioningRequestProcessor for Check-capacity ProvisioningClass.
+func NewCheckCapacityProcessor(kubeConfig *rest.Config) (*checkCapacityProcessor, error) {
+	client, err := provreqclient.NewProvisioningRequestClient(kubeConfig)
+	if err != nil {
+		return nil, err
+	}
+	return &checkCapacityProcessor{client: client, now: time.Now}, nil
+}
+
+// Process iterates over ProvisioningRequests and apply:
+// -BookingExpired condition for Provisioned ProvisioningRequest if capacity reservation time is expired.
+// -Failed condition for ProvisioningRequest that were not provisioned during defaultExpirationTime.
+// TODO(yaroslava): fetch reservation and expiration time from ProvisioningRequest
+func (p *checkCapacityProcessor) Process() error {
+	provReqs, err := p.client.ProvisioningRequests()
+	if err != nil {
+		return fmt.Errorf("Couldn't fetch ProvisioningRequests in the cluster: %v", err)
+	}
```

Comment:
Shouldn't we cap the number of ProvREqs we handle in one loop to not spend a lot of time in this refresh?
Can we start with 10-20 ProvReqs?

Reply:
I don't think we should limit the number of ProvReqs, we need to make sure we have updated conditions for all ProvReq

Reply:
Ack, I would still prefer to have some cap to avoid overloading CA in this part and starving the actual scale up logic.

Reply:
We might spend some time here, but this won't happen each CA loop, because we setCondition once for each ProvReq: or BookingExpired or Failed at some point. 
I'm worried that with ProvReqs cut, we might have a situation when some ProvReq won't be processed for sometime, so we overbook capacity for it or tried to scaleUp that will result in blocking other ProvReqs from scaleUp.

Reply:
make sense, updated